Framework/Algorithms/src/SumSpectra.cpp

```diff
@@ -222,69 +240,73 @@ SumSpectra::getOutputSpecNo(MatrixWorkspace_const_sptr localworkspace) {
  * @param numZeros The number of zero bins in histogram workspace or empty
  * spectra for event workspace.
  */
-void SumSpectra::doWorkspace2D(MatrixWorkspace_const_sptr localworkspace,
-                               ISpectrum &outSpec, Progress &progress,
+void SumSpectra::doWorkspace2D(ISpectrum &outSpec, Progress &progress,
                                size_t &numSpectra, size_t &numMasked,
                                size_t &numZeros) {
   // Get references to the output workspaces's data vectors
-  auto &YSum = outSpec.mutableY();
-  auto &YError = outSpec.mutableE();
+  auto &OutputYSum = outSpec.mutableY();
+  auto &OutputYError = outSpec.mutableE();
 
   std::vector<double> Weight;
   std::vector<size_t> nZeros;
   if (m_calculateWeightedSum) {
-    Weight.assign(YSum.size(), 0);
-    nZeros.assign(YSum.size(), 0);
+    Weight.assign(OutputYSum.size(), 0);
+    nZeros.assign(OutputYSum.size(), 0);
   }
   numSpectra = 0;
   numMasked = 0;
   numZeros = 0;
 
+  MatrixWorkspace_sptr in_ws = getProperty("InputWorkspace");
+  // Clean workspace of any NANs or Inf values
+  auto localworkspace = replaceSpecialValues(in_ws);
   const auto &spectrumInfo = localworkspace->spectrumInfo();
   // Loop over spectra
-  for (const auto i : this->m_indices) {
+  for (const auto wsIndex : this->m_indices) {
     // Don't go outside the range.
-    if ((i >= this->m_numberOfSpectra) || (i < 0)) {
-      g_log.error() << "Invalid index " << i
+    if ((wsIndex >= this->m_numberOfSpectra) || (wsIndex < 0)) {
+      g_log.error() << "Invalid index " << wsIndex
                     << " was specified. Sum was aborted.\n";
       break;
     }
 
-    if (spectrumInfo.hasDetectors(i)) {
+    if (spectrumInfo.hasDetectors(wsIndex)) {
       // Skip monitors, if the property is set to do so
-      if (!m_keepMonitors && spectrumInfo.isMonitor(i))
+      if (!m_keepMonitors && spectrumInfo.isMonitor(wsIndex))
         continue;
       // Skip masked detectors
-      if (spectrumInfo.isMasked(i)) {
+      if (spectrumInfo.isMasked(wsIndex)) {
         numMasked++;
         continue;
       }
     }
     numSpectra++;
 
+    const auto &YValues = localworkspace->y(wsIndex);
+    const auto &YErrors = localworkspace->e(wsIndex);
+
     // Retrieve the spectrum into a vector
-    const auto &YValues = localworkspace->y(i);
-    const auto &YErrors = localworkspace->e(i);
-    if (m_calculateWeightedSum) {
-      for (int k = 0; k < this->m_yLength; ++k) {
-        if (YErrors[k] != 0) {
-          double errsq = YErrors[k] * YErrors[k];
-          YError[k] += errsq;
-          Weight[k] += 1. / errsq;
-          YSum[k] += YValues[k] / errsq;
+
+    for (size_t i = 0; i < YValues.size(); ++i) {
```

Comment:
Why change from `this->m_yLength`? The `this->m_yLength` is used everywhere else, so I think it might be confusing as to why this single instance is `YValues.size()`, even if they have the same size